File: aoc-python/day02/part2.py
from __future__ import annotations

import argparse
import os.path

import pytest

import support

INPUT_TXT = os.path.join(os.path.dirname(__file__), 'input.txt')

    
def rps(elf, strategy):
    score_map = {
        "A": 1,
        "B": 2,
        "C": 3,
    }
    rps_win = {
        "A": "B",
        "B": "C",
        "C": "A",
    }
    rps_lose = {v: k for k, v in rps_win.items()}

    if strategy == "X":
        score = 0 + score_map[rps_lose[elf]]
    elif strategy == "Y":
        score = 3 + score_map[elf]
    else:
        score = 6 + score_map[rps_win[elf]]

    return score


def compute(s: str) -> int:
    lines = s.splitlines()
    score = 0
    for line in lines:
        print(line)
        score += rps(line[0], line[2])
    return score


INPUT_S = '''\
A Y
B X
C Z
'''
EXPECTED = 12


@pytest.mark.parametrize(
    ('input_s', 'expected'),
    (
        (INPUT_S, EXPECTED),
    ),
)
def test(input_s: str, expected: int) -> None:
    assert compute(input_s) == expected


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument('data_file', nargs='?', default=INPUT_TXT)
    args = parser.parse_args()

    with open(args.data_file) as f, support.timing():
        print(compute(f.read()))

    return 0


if __name__ == '__main__':
    raise SystemExit(main())

File: aocJulia/days/day02/part2.jl
using Test

using aocJulia: helpers


function compute(s)
    games = split.(split(chomp(s), "\n"), " ")
    score_map = Dict(
        "A" => 1,
        "B" => 2,
        "C" => 3
    )
    win_map = Dict(
        "A" => "B",
        "B" => "C",
        "C" => "A"
    )
    lose_map =  Dict(values(win_map) .=> keys(win_map))

    score = 0
    for (elf, strategy) in games
        if strategy == "Z"
            score += 6 + score_map[win_map[elf]]
        elseif strategy == "X"
            score += score_map[lose_map[elf]]
        else
            score += 3 + score_map[elf]
        end
    end

    return score
end


INPUTS_S = """
A Y
B X
C Z
"""
EXPECTED = 12


@test compute(INPUTS_S) == EXPECTED

INPUT_TXT = helpers.read_input(dirname(Base.source_path()))
println(compute(INPUT_TXT))